### PR TITLE
Remove temporary files not deleted by destructor on failure

### LINKED ISSUE #2022: PHP-CS-Fixer leaves a temporary file at the root of the project
When using the 1.11.4 version, a ``cs_fixer_tmp_...`` file stays at the root of the project at the end of the run

### PR DESCRIPTION
**Pull request by adawolfa**
This PR should fix #2022

LintManager creates temporary files and deletes them in destructor. In certain circumstances (i. e. fatal error), the destructor not get called. I have solved this by registering shutdown function, which unlinks files left in index.

I'm aware this is not _clean_ solution and atm I have no idea how to write utest for this behavior. Any suggestions are welcome.

### CODE REVIEW

File: Symfony/CS/LintManager.php [hunk 1]
```diff
@@ -67,11 +83,22 @@ public function createProcessForSource($source)
     {
         if (null === $this->temporaryFile) {
             $this->temporaryFile = tempnam('.', 'cs_fixer_tmp_');
+            self::$temporaryFiles[$this->temporaryFile] = $this->temporaryFile;
```
Comment (keradus) [inline Symfony/CS/LintManager.php]: `[foo => foo, bar => bar]` :(

Comment (adawolfa) [inline Symfony/CS/LintManager.php]: Should I replace value by a TRUE, or should I use in_array?

**Comment by keradus** [comment]
Nice work so far!

I would move that logic to separated class.
sth like
```php
interface ShutdownFileRemoval
{
    public function add(string $file);
    public function remove(string $file);
    public function remove();
}
```

Each instance of `ShutdownFileRemoval` would register it's own shutdown function.

Then, you will have no static methods. A lot easier to test.

Also, thanks to this, you will be apply to apply same logic in `AbstractIntegrationTest` about `$tmpFile`

File: Symfony/CS/LintManager.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/LintManager.php]: check out my last proposal

**Comment by adawolfa** [comment]
OK, stay tuned.

File: Symfony/CS/ShutdownFileRemoval.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * Removes a file from shutdown removal.
+     *
+     * @param string $path
+     */
+    public function detach($path)
+    {
+        if ($key = array_search($path, $this->files, true)) {
```
Comment (keradus) [inline Symfony/CS/ShutdownFileRemoval.php]: do not assign inside condition

File: Symfony/CS/ShutdownFileRemoval.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    }
+
+    /**
+     * Removes attached files.
+     */
+    public function clean()
+    {
+        foreach ($this->files as $file) {
+            @unlink($file); // @ - intentionally
```
Comment (keradus) [inline Symfony/CS/ShutdownFileRemoval.php]: then describe why it's intentionally

File: Symfony/CS/ShutdownFileRemoval.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+namespace Symfony\CS;
+
+/**
+ * Handles left over temporary files removal.
+ *
+ * @author Adam Klvač <adam@klva.cz>
+ */
+class ShutdownFileRemoval
```
Comment (keradus) [inline Symfony/CS/ShutdownFileRemoval.php]: final + `@internal`

**Comment by keradus** [comment]
Please update integration test as well ;)

**Comment by keradus** [comment]
Thank you @adawolfa.

**Pull request closed without merging**